### PR TITLE
Add timeout for waiting on compactor to become ACTIVE in the ring.

### LINKED ISSUE #3914: compactor went into tailspin
**Describe the bug**

Consul was down at the time compactor started, and it never recovered:

```
level=info ts=2021-03-04T20:23:01.350819279Z caller=main.go:188 msg="Starting Cortex" version="(version=1.6.0, branch=master, revision=56f794d)"
level=info ts=2021-03-04T20:23:01.35627101Z caller=module_service.go:59 msg=initialising module=server
level=info ts=2021-03-04T20:23:01.358032071Z caller=module_service.go:59 msg=initialising module=compactor
level=info ts=2021-03-04T20:23:01.351964058Z caller=server.go:229 http=[::]:80 grpc=[::]:9095 msg="server listening on addresses"
level=info ts=2021-03-04T20:23:01.356491008Z caller=module_service.go:59 msg=initialising module=memberlist-kv
level=info ts=2021-03-04T20:23:01.36340712Z caller=compactor.go:373 component=compactor msg="waiting until compactor is ACTIVE in the ring"
level=info ts=2021-03-04T20:23:01.366220882Z caller=lifecycler.go:527 msg="not loading tokens from file, tokens file path is empty"
level=error ts=2021-03-04T20:23:01.391200499Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.394262836Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.391467754Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.399955867Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.397642587Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.402454173Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.404896477Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.407444107Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.410901197Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.413247689Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:01.4157827Z caller=client.go:147 msg="error getting key" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:02.848432818Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:06.895736201Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:14.397161217Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:25.078769837Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:23:49.041899802Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:24:35.634427793Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:25:35.127795679Z caller=client.go:234 msg="error getting path" key=compactor err="Get \"http://consul.cortex.svc.cluster.local:8500/v1/kv/compactor?stale=&wait=10000ms\": dial tcp: lookup consul.cortex.svc.cluster.local on 10.0.0.10:53: no such host"
level=error ts=2021-03-04T20:26:37.176005708Z caller=client.go:234 msg="error getting path" key=compactor err="Unexpected response code: 500"
level=error ts=2021-03-04T20:27:42.508689928Z caller=client.go:234 msg="error getting path" key=compactor err="Unexpected response code: 500"
level=error ts=2021-03-04T20:28:35.350979964Z caller=client.go:234 msg="error getting path" key=compactor err="Unexpected response code: 500"
level=error ts=2021-03-04T20:29:41.117751261Z caller=client.go:234 msg="error getting path" key=compactor err="Unexpected response code: 500"
level=error ts=2021-03-04T20:30:39.87687265Z caller=client.go:234 msg="error getting path" key=compactor err="Unexpected response code: 500"
level=info ts=2021-03-04T20:31:36.036454597Z caller=client.go:247 msg="value is nil" key=compactor index=22
level=info ts=2021-03-04T20:31:36.930199939Z caller=client.go:247 msg="value is nil" key=compactor index=24
level=info ts=2021-03-04T20:31:37.932395975Z caller=client.go:247 msg="value is nil" key=compactor index=25
level=info ts=2021-03-04T20:31:41.887641219Z caller=client.go:247 msg="value is nil" key=compactor index=28
level=info ts=2021-03-04T20:31:41.941646251Z caller=client.go:247 msg="value is nil" key=compactor index=29
level=info ts=2021-03-04T20:31:46.926493102Z caller=client.go:247 msg="value is nil" key=compactor index=31
level=info ts=2021-03-04T20:31:46.967013886Z caller=client.go:247 msg="value is nil" key=compactor index=32
...
level=info ts=2021-03-05T10:12:35.156472897Z caller=client.go:247 msg="value is nil" key=compactor index=78733
level=info ts=2021-03-05T10:12:35.455228914Z caller=client.go:247 msg="value is nil" key=compactor index=78734
level=info ts=2021-03-05T10:12:36.157313862Z caller=client.go:247 msg="value is nil" key=compactor index=78736
level=info ts=2021-03-05T10:12:37.157305834Z caller=client.go:247 msg="value is nil" key=compactor index=78738
```

We have compactor sharding turned on:
```
        - -compactor.ring.consul.hostname=consul.cortex.svc.cluster.local:8500
        - -compactor.ring.prefix=
        - -compactor.ring.store=consul
        - -compactor.sharding-enabled=true
```

**Expected behavior**
I think it should exit with error in this situation; crashlooping would make the fault more obvious to the operator, and after a few restarts it would have managed to talk to Consul in my case.

### PR DESCRIPTION
**Pull request by ac1214**
Signed-off-by: Albert <ac1214@users.noreply.github.com>

**What this PR does**:
Addresses #3914
**Which issue(s) this PR fixes**:
Fixes #3914

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/compactor/compactor.go
Comment (bboreham) [inline pkg/compactor/compactor.go:46]: This seems weird. `time.Duration(600)` is 600ns.
I recommend you make this `600*time.Second` (or perhaps `10*time.Minute`) and remove the multiplication on line 398.

Also why isn't it a constant?  (Maybe it should be configurable, but you didn't do that)

File: pkg/ring/ring.go
Comment (bboreham) [inline pkg/ring/ring.go:770]: It would be ok to leave the description here.

**Review by bboreham — COMMENTED** [review]
Thanks for the PR! It's broadly ok, but a couple of style points.

It looks to me like the tests you've added test similar code, rather than the code you added. If I'm right, could you test the actual code?

What is the user experience if the timeout is hit?  Do they see anything in the logs to let them know what happened?

File: pkg/compactor/compactor.go
Comment (ac1214) [reply pkg/compactor/compactor.go]: Thanks for the review, I made changes to allow the timeout to be configurable.

**Comment by ac1214** [comment]
> Thanks for the PR! It's broadly ok, but a couple of style points.
> 
> It looks to me like the tests you've added test similar code, rather than the code you added. If I'm right, could you test the actual code?
> 
> What is the user experience if the timeout is hit? Do they see anything in the logs to let them know what happened?

I added tests to actually test the starting of the compactor. I also added some logs as to make it clear to why the compactor failed to start.

**Comment by bboreham** [comment]
Sorry but we have begun the process of cutting a new release; please rebase from `master` and move your CHANGELOG entry to the top under `## master / unreleased`

File: pkg/compactor/compactor.go
Comment (pracucci) [inline pkg/compactor/compactor.go:397]: We shouldn't overwrite `ctx` otherwise the timeout applies to next steps too. You should name the returned context variable differently and use it below where you want the timeout to be applied.

File: pkg/compactor/compactor_ring.go
Comment (pracucci) [inline pkg/compactor/compactor_ring.go:66]: [nit] Period at the end.

```suggestion
	f.DurationVar(&cfg.StartingTimeout, "compactor.ring.starting-timeout", 10*time.Minute, "Timeout for waiting on compactor to become ACTIVE in the ring.")
```

Comment (pracucci) [inline pkg/compactor/compactor_ring.go:66]: Sorry for being picky. I would make the CLI flag (and the YAML config option) name even more explicit:
- CLI flag: `compactor.ring.wait-active-instance-timeout`
- YAML field: `wait_active_instance_timeout`

File: pkg/compactor/compactor_test.go
Comment (pracucci) [inline pkg/compactor/compactor_test.go:1303]: [nit]
```suggestion
	// Assert that the compactor timed out
```

File: pkg/ring/util_test.go
Comment (pracucci) [inline pkg/ring/util_test.go:248]: I guess you wanted 3s and not 3ns. You should change this to `time.Duration(3 * time.Second)`. I think 1s is enough for this test, so I would just do `time.Second`.

_Same comment applies to other tests._

**Review by pracucci — COMMENTED** [review]
Good job and thanks for addressing my feedback. Way easier than the initial version. I left few last comments 🙏

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:6]: [nit] We use CHANGE for breaking changes. This is more an enhancement.

```suggestion
* [ENHANCEMENT] Add timeout for waiting on compactor to become ACTIVE in the ring. #4262
```

**Review by pracucci — APPROVED** [review]
LGTM (modulo a nit)

File: CHANGELOG.md
Comment (ac1214) [reply CHANGELOG.md]: Fixed! Thanks for the review!

**Comment by ac1214** [comment]
Would I be able to get a second review on this PR and get the workflow triggered again? @bboreham
Thanks!

**Review by bboreham — APPROVED** [review]
LGTM

**Pull request merged**
